graph/breadth_first_search.cpp

```diff
@@ -0,0 +1,135 @@
+ * are the number of edges in the graph.
+ *
+ * Applications of Breadth First Search are
+ *
+ * 1. Finding shortest path between two vertices say u and v, with path
+ *    length measured by number of edges (an advantage over depth first
+ *    search algorithm)
+ * 2. Ford-Fulkerson Method for computing the maximum flow in a flow network.
+ * 3. Testing bipartiteness of a graph.
+ * 4. Cheney's Algorithm, Copying garbage collection.
+ *
+ * And there are many more...
+ *
+ * <h4>working</h4>
+ * In the implementation below we first created a graph using the adjacency
+ * list representation of graph.
+ * Breadth First Search Works as follows
+ * it requires a vertex as a start vertex, Start vertex is that vertex
+ * from where you want to start traversing the graph.
+ * we maintain a bool array or a vector to keep track of the vertices
+ * which we have visited so that we do not traverse the visited vertices
+ * again and again and eventually fall into an infinite loop. Along with this
+ * boolen array we use a Queue.
+ *
+ * 1. First we mark the start vertex as visited.
+ * 2. Push this visited vertex in the Queue.
+ * 3. while the queue is not empty we repeat the following steps
+ *
+ *      1. Take out an element from the front of queue
+ *      2. start exploring the adjacency list of this vertex
+ *         if element in the adjacency list is not visited then we
+ *         push that element into the queue and mark this as visited
+ *
+ */
+#include <iostream>
+#include <queue>
+#include <vector>
+
+/**
+ * \namespace graph
+ * \brief Graph algorithms
+ */
+namespace graph {
+/**
+ * \brief
+ * Adds and edge between two vertices of graph say u and v in this
+ * case.
+ *
+ * @param adj Adjacency list representation of graph
+ * @param u first vertex
+ * @param v second vertex
+ *
+ */
+void addEdge(std::vector<std::vector<int>> *adj, size_t u, size_t v) {
+    /**
+     * Here we are considering directed graph that's the
+     * reason we are adding v to the adjacency list representation of u
+     * but not adding u to the adjacency list representation of v
+     *
+     * in case of a un-directed graph you can un comment the statement below.
+     */
+    (*adj)[u - 1].push_back(v - 1);
+    // adj[v - 1].push_back(u -1);
+}
+
+/**
+ * \brief
+ * Function performs the breadth first search algorithm over the graph
+ *
+ * @param adj Adjacency list representation of graph
+ * @param start vertex from where traversing starts
+ *
+ */
+void beadth_first_search(std::vector<std::vector<int>> *adj, size_t start) {
+    size_t vertices = adj->size();
+
+    /// vector to keep track of visited vertices
+    std::vector<bool> visited(vertices, 0);
+
+    std::queue<int> tracker;
+    /// marking the start vertex as visited
+    visited[start] = true;
+    tracker.push(start);
+    while (!tracker.empty()) {
+        size_t vertex = tracker.front();
+        tracker.pop();
+        std::cout << vertex + 1 << " ";
```

Comment:
It would be better if the function is made modular by returning the result instead of printing it.
The return value can then be utilized by a test function and make the code more manageable

Reply:
yes you are right but i had not done this beacuse the purpose of these files is only to demonstrate the traversal techinque used to traverse the graphs.

another reason to not to return the value is if i return the value `v+1` from here then the control of program is transfered to the calling function that is the main funcition in this case and one the control is transfered to main function without completely executing the algorithm we cannot get back to the ```breadth_first_search()``` function.

for test ulility what we can do is that we can push these values into a vector or in some other similar structure and then we can test this vector in our tests, but for this too either we need to pass an extra vector by reference as the function argument or we have to maintain a vector globally.

that's why i dropped the idea of writing tests in these traversal algorithms.

Reply:
got it. how about returning an array instead? I am pretty sure traversing a tree can be beneficial in robotics, for example

Reply:
yes not only in robotics, infact there are a lot of applications of these traversal algorithms.

let's suppose if we return an array there could be two possible approches
- we can pass an array in the breadth_first_search() funciton to fill it with required data and then return this
- and other one is we can maintain a global array then there is no need to returning any array

i will favour the second point because if we choose this then this will not modify the function declaration that is `void breadth_first_seach(parameters...)` and this seems reasonable as the algorithm itself does not return any value it just traverses the graph and keeps the testing part separate but in the first approch we need to modify the function header to return an `int*` pointing to array and it might be a point of confusion for readers.

Reply:
@kvedala what's your opinion?

Reply:
instantiate a local variable of type `std::vector<T>`.
In this loop, the resulting elements can be appended to it and at the end can be returned. The `std::vector` and `std::array` classes have copy contracts and hence can be safely assigned to in the calling function.

Reply:
> instantiate a local variable of type `std::vector<T>`.
> In this loop, the resulting elements can be appended to it and at the end can be returned. The `std::vector` and `std::array` classes have copy contracts and hence can be safely assigned to in the calling function.
> Thus, any global variable can be avoided

okay i was thinking that this local variable will get destroyed as soon as the breadth_first_search function smashes from the calling stack.
i was not aware of this that this is the case with c-style arrays.

Reply:
Just to confirm, what I am suggesting is something like so:
```cpp
std::vector<T> beadth_first_search(std::vector<std::vector<int>> *adj, size_t start) {
```
if the resultant size is unknown. If the resulting size is known, then
```cpp
std::array<T, N> beadth_first_search(std::vector<std::vector<int>> *adj, size_t start) {
```
will be more efficient

Reply:
resultant size is unknown because we are taking input from the used so i used ```std::vector```
@kvedala have a look at the implementaion